pinot-core/src/main/java/org/apache/pinot/core/common/DataSourceMetadata.java

```diff
@@ -18,24 +18,79 @@
  */
 package org.apache.pinot.core.common;
 
-import org.apache.pinot.spi.data.FieldSpec.DataType;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.pinot.core.data.partition.PartitionFunction;
+import org.apache.pinot.spi.data.FieldSpec;
 
 
+/**
+ * The {@code DataSourceMetadata} class contains all the metadata for a column for query execution purpose.
+ */
 public interface DataSourceMetadata {
 
-  DataType getDataType();
+  /**
+   * Returns the field spec of the column.
+   */
+  FieldSpec getFieldSpec();
+
+  /**
+   * Returns the data type of the column.
+   */
+  default FieldSpec.DataType getDataType() {
+    return getFieldSpec().getDataType();
+  }
 
-  boolean isSingleValue();
+  /**
+   * Returns {@code true} if the column is single-valued, {@code false} otherwise.
+   */
+  default boolean isSingleValue() {
+    return getFieldSpec().isSingleValueField();
+  }
 
+  /**
+   * Returns {@code true} if the column is sorted, {@code false} otherwise.
+   */
   boolean isSorted();
 
+  /**
+   * Returns the number of documents of the column.
+   */
   int getNumDocs();
 
-  int getMaxNumMultiValues();
+  /**
+   * Returns the number of values of the column. For single-value column, number of values always equals to number of
+   * documents; for multi-value column, each document (multi-value entry) may contain multiple values.
+   */
+  int getNumValues();
+
+  /**
+   * Returns the maximum number of values for each multi-value entry of the multi-value column. For single-value column,
+   * the value is undefined.
```

Comment:
might as well throw exception instead of returning an undefined value and getting an undefined behavior?

Reply:
Want to avoid the if check for every call. Make it return -1 for single-value column